[conf_handling] New ability - modify configuration line for freshclam & clamd

ClementJ35 opened the pull request:
# Ability to manage freshclam and clamd configuration file #

## 🗣 Description ##

This PR adds the ability to construct a dictionary of values to be integrated in both the freshclam and clamd configuration files.

## 💭 Motivation and context ##

I need to handle modifications to these configurations with Ansible.

## 🧪 Testing ##

Automated testing passes.  Tested independently on Debian 11.

## ✅ Pre-approval checklist ##

<!-- Remove any of the following that do not apply. -->
<!-- Draft PRs should have one or more unchecked boxes. -->
<!-- If you're unsure about any of these, don't hesitate to ask. -->
<!-- We're here to help! -->

- [x] This PR has an informative and human-readable title.
- [x] Changes are limited to a single goal - *eschew scope creep!*
- [x] *All* future TODOs are captured in issues, which are referenced
      in code comments.
- [x] All relevant type-of-change labels have been added.
- [x] I have read the [CONTRIBUTING](../blob/develop/CONTRIBUTING.md) document.
- [x] These code changes follow [cisagov code standards](https://github.com/cisagov/development-guide).
- [x] All relevant repo and/or project documentation has been updated
      to reflect the changes in this PR.
- [x] Tests have been added and/or modified to cover the changes in this PR.
- [x] All new and existing tests pass.

ClementJ35 committed:
[conf_handling] One time commit for new ability of conf handling

ClementJ35 committed:
Fixing linting problems

ClementJ35 committed:
[configure] Add a way to remove entirely a key from conf file

ClementJ35 committed:
[linting] fix readme

ClementJ35 committed:
[configure] delete debug task

ClementJ35 committed:
[configure] Secure modified configuration file to 0700

It's required with some configuration line like HTTPProxyPassword

jsf9k commented on README.md:
```diff
@@ -19,6 +19,39 @@ None.
 | clamav_cron_frequency | The frequency of ClamAV scanning.  Must be one of: `hourly`, `daily`, `weekly`, or `monthly`. | `weekly` | No |
+| clamav_freshclam_conf | Dictionnary of values to set in freshclam configuration file  | false | No |
+| clamav_clamd_conf | Dictionnary of values to set in clamd configuration file  | false | No |
+| clamav_configuration_backup | Shall we backup configuration files when changed | false | No |
```
Clean up the language and grammar.

```suggestion
| clamav_freshclam_conf | A dictionary of values to set in the freshclam configuration file. | false | No |
| clamav_clamd_conf | A dictionary of values to set in the clamd configuration file. | false | No |
| clamav_configuration_backup | Whether or not to backup configuration files before changing. | false | No |
```

jsf9k commented on defaults/main.yml:
```diff
@@ -1 +1,20 @@
+# Dictionnary of values to be configured.
+# Multivalued key are described in list.
+# None value means : clear the Key entirely from the file.
+# Example :
```
```suggestion
# A dictionary of values to be configured in the clamd and freshclam configuration files.
# Multivalued keys are specified as lists.
# A value of None means that the key will be removed from the file.
# Example:
```

jsf9k commented on tasks/main.yml:
```diff
@@ -35,6 +35,20 @@
+  - task_conf_parameters: "{{ clamav_clamd_conf }}"
+  ansible.builtin.include_tasks: configure.yml
+  when: clamav_clamd_conf
+
```
Move the task so that it comes before the `vars`.

```suggestion
- name: Ensure configuration for freshclam
  ansible.builtin.include_tasks: configure.yml
  vars:
  - task_conf_file: "{{ freshclam_configuration_path }}"
  - task_conf_parameters: "{{ clamav_freshclam_conf }}"
  when: clamav_freshclam_conf

- name: Ensure configuration for clamd
  ansible.builtin.include_tasks: configure.yml
  vars:
  - task_conf_file: "{{ clamd_configuration_path }}"
  - task_conf_parameters: "{{ clamav_clamd_conf }}"
  when: clamav_clamd_conf

```

jsf9k commented on README.md:
````diff
@@ -19,6 +19,39 @@ None.
+  PrivateMirror: None
+```
+
+would change :
````
Remove unnecessary space.

```suggestion
would change:
```

jsf9k commented on README.md:
````diff
@@ -19,6 +19,39 @@ None.
+
+to :
+
+```shell
````
I think this will look better with `properties` syntax highlighting:
```suggestion
```properties
```

jsf9k commented on README.md:
````diff
@@ -19,6 +19,39 @@ None.
+
+would change :
+
+```shell
````
I think this will look better with `properties` syntax highlighting:
```suggestion
```properties
```

jsf9k commented on tasks/configure.yml:
```diff
@@ -0,0 +1,36 @@
+---
+- name: "Configure | Single-valued - Ensure value exists in {{ task_conf_file }}"
```
```suggestion
- name: Configure single-valued entries in {{ task_conf_file }}
```

jsf9k commented on tasks/configure.yml:
```diff
@@ -0,0 +1,36 @@
+    state: present
+  loop: "{{ task_conf_parameters | dict2items | selectattr('value','string') }}"
+
+- name: "Configure | Multi-valued - Ensure values exists in {{ task_conf_file }}"
```
```suggestion
- name: Configure multi-valued entries in {{ task_conf_file }}
```

jsf9k commented on tasks/configure.yml:
```diff
@@ -0,0 +1,36 @@
+    state: present
+  loop: "{{ task_conf_parameters | dict2items | rejectattr('value','string') | subelements( 'value' ) }}"
+
+- name: "Configure | Remove unwanted values in {{ task_conf_file }}"
```
```suggestion
- name: Remove unwanted values from {{ task_conf_file }}
```

jsf9k commented on tasks/configure.yml:
```diff
@@ -0,0 +1,36 @@
+    backup: "{{ clamav_configuration_backup }}"
+  with_items: "{{ task_conf_parameters }}"
+
+- name: "Configure | Clear key with None value"
```
```suggestion
- name: Remove unwanted keys from {{ task_conf_file }}
```

jsf9k commented on tasks/configure.yml:
```diff
@@ -0,0 +1,36 @@
+---
+- name: "Configure | Single-valued - Ensure value exists in {{ task_conf_file }}"
+  lineinfile:
```
Use FQCNs:
```suggestion
  ansible.builtin.lineinfile:
```

jsf9k commented on tasks/configure.yml:
```diff
@@ -0,0 +1,36 @@
+  loop: "{{ task_conf_parameters | dict2items | selectattr('value','string') }}"
+
+- name: "Configure | Multi-valued - Ensure values exists in {{ task_conf_file }}"
+  lineinfile:
```
Use FQCNs:
```suggestion
  ansible.builtin.lineinfile:
```

jsf9k commented on tasks/configure.yml:
```diff
@@ -0,0 +1,36 @@
+  loop: "{{ task_conf_parameters | dict2items | rejectattr('value','string') | subelements( 'value' ) }}"
+
+- name: "Configure | Remove unwanted values in {{ task_conf_file }}"
+  lineinfile:
```
Use FQCNs:
```suggestion
  ansible.builtin.lineinfile:
```

jsf9k commented on tasks/configure.yml:
```diff
@@ -0,0 +1,36 @@
+  with_items: "{{ task_conf_parameters }}"
+
+- name: "Configure | Clear key with None value"
+  lineinfile:
```
Use FQCNs:
```suggestion
  ansible.builtin.lineinfile:
```

jsf9k commented on tasks/configure.yml:
```diff
@@ -0,0 +1,36 @@
+  loop: "{{ task_conf_parameters | dict2items | selectattr('value','none') }}"
+
+- name: "Configure | Secure configuration files"
+  file:
```
Use FQCNs:
```suggestion
  ansible.builtin.file:
```

jsf9k commented on tasks/configure.yml:
```diff
@@ -0,0 +1,36 @@
+    backup: "{{ clamav_configuration_backup }}"
+  loop: "{{ task_conf_parameters | dict2items | selectattr('value','none') }}"
+
+- name: "Configure | Secure configuration files"
```
```suggestion
- name: Set permissions on configuration file {{ task_conf_file }}
```

jsf9k commented on README.md:
````diff
@@ -19,6 +19,39 @@ None.
+  ...
+```
+
+to :
````
Remove unnecessary space:
```suggestion
to:
```

jsf9k reviewed (changes requested):
Thanks for creating another PR @ClementJ35!

In addition to the changes I suggested below, I'd like to see some `molecule` tests added to cover the tasks in the `configure.yml` file.  You should create one or more new `molecule` scenarios for this, as is done, e.g., in cisagov/ansible-role-python#46.  (There the `python2` scenario was created as an additional `molecule` scenario.)

ClementJ35 committed:
Apply suggestions from code review

Those are all good suggestions.

Co-authored-by: Shane Frasier <maverick@maverickdolphin.com>

ClementJ35 commented:
Hi @jsf9k,
Happy to help.
I'll look into molecule testing and come back with commits.

ClementJ35 committed:
[syntax] change vars injection list to a dict in main task

ClementJ35 committed:
[configure] Change backing up plan to a dedicated task, it makes more sense

ClementJ35 committed:
[configure] Better handling of None value

- Switch with_items for loop
- Filter the loop entries
- Bugfix for loop syntax

ClementJ35 committed:
[lint] fix EOF

ClementJ35 committed:
[test] Add configuration variables in role play to test

ClementJ35 committed:
[test] first try

ClementJ35 committed:
[test] Correct paste var + delete pytest import

ClementJ35 committed:
[test] Add docstring

ClementJ35 committed:
[test] dict in python style, not yaml

ClementJ35 committed:
[test] fix flake8 checks

ClementJ35 committed:
[test] pass black linter

ClementJ35 committed:
[test] change file openging and add some docstring

ClementJ35 committed:
[test] rollback file handling

ClementJ35 committed:
[test] Debug : show file_content output

ClementJ35 committed:
[test] delete debug and fix path in rhel-like OS

ClementJ35 committed:
[test] and fix path in test too

ClementJ35 committed:
[configure] secure file permission on backup

ClementJ35 committed:
[test] pass debug to molecule see what's really tested

ClementJ35 committed:
[test] ensure all configuration entries were parsed

ClementJ35 committed:
[lint] fix whitespace

ClementJ35 committed:
[test] len() is not a list method...

ClementJ35 committed:
[test] splitlines from content_string

ClementJ35 committed:
[test] fix forgotten definition

ClementJ35 committed:
[test] fix typo in converge.yml

ClementJ35 committed:
[test] deactivate debug switch for molecule

ClementJ35 commented:
@jsf9k , I added test in [test_configure.py](molecule/tests/test_configure.py).
It's now ready to review.

ClementJ35 committed:
Merge branch 'develop' into conf_handling

jsf9k commented on tasks/configure.yml:
```diff
@@ -0,0 +1,44 @@
+---
+- name: Backup {{ task_conf_file }} to {{ task_conf_file }}.{{ ansible_date_time.iso8601 }}
+  ansible.builtin.copy:
+    src: "{{ task_conf_file }}"
+    dest: "{{ task_conf_file }}.{{ ansible_date_time.iso8601 }}"
+    mode: 0700
+    remote_src: true
+  when: clamav_configuration_backup
+  tags:
+    - molecule-idempotence-notest
```
Alphabetize:
```suggestion
- name: Backup {{ task_conf_file }} to {{ task_conf_file }}.{{ ansible_date_time.iso8601 }}
  ansible.builtin.copy:
    dest: "{{ task_conf_file }}.{{ ansible_date_time.iso8601 }}"
    mode: 0700
    remote_src: true
    src: "{{ task_conf_file }}"
  tags:
    - molecule-idempotence-notest
  when: clamav_configuration_backup
```

jsf9k commented on tasks/configure.yml:
```diff
@@ -0,0 +1,44 @@
+  tags:
+    - molecule-idempotence-notest
+
+- name: Configure single-valued entries in {{ task_conf_file }}
+  ansible.builtin.lineinfile:
+    path: "{{ task_conf_file }}"
+    line: "{{ item.key }} {{ item.value }}"
+    search_string: "{{ item.key }} "
+    state: present
+  loop: "{{ task_conf_parameters | dict2items | selectattr('value','string') }}"
```
Alphabetize:
```suggestion
- name: Configure single-valued entries in {{ task_conf_file }}
  ansible.builtin.lineinfile:
    line: "{{ item.key }} {{ item.value }}"
    path: "{{ task_conf_file }}"
    search_string: "{{ item.key }} "
    state: present
  loop: "{{ task_conf_parameters | dict2items | selectattr('value','string') }}"
```

jsf9k commented on tasks/configure.yml:
```diff
@@ -0,0 +1,44 @@
+    state: present
+  loop: "{{ task_conf_parameters | dict2items | selectattr('value','string') }}"
+
+- name: Configure multi-valued entries in {{ task_conf_file }}
+  ansible.builtin.lineinfile:
+    path: "{{ task_conf_file }}"
+    line: "{{ item.0.key }} {{ item.1 }}"
+    state: present
+  loop: "{{ task_conf_parameters | dict2items | rejectattr('value','string') | rejectattr('value','none') | subelements( 'value' ) }}"
```
Alphabetize:
```suggestion
- name: Configure multi-valued entries in {{ task_conf_file }}
  ansible.builtin.lineinfile:
    line: "{{ item.0.key }} {{ item.1 }}"
    path: "{{ task_conf_file }}"
    state: present
  loop: "{{ task_conf_parameters | dict2items | rejectattr('value','string') | rejectattr('value','none') | subelements( 'value' ) }}"
```

jsf9k commented on tasks/configure.yml:
```diff
@@ -0,0 +1,44 @@
+    state: absent
+  loop: "{{ task_conf_parameters | dict2items | selectattr('value','none') }}"
+
+- name: Set permissions on configuration file {{ task_conf_file }}
+  ansible.builtin.file:
+    path: "{{ task_conf_file }}"
+    mode: 0700
```
Alphabetize:
```suggestion
- name: Set permissions on configuration file {{ task_conf_file }}
  ansible.builtin.file:
    mode: 0700
    path: "{{ task_conf_file }}"
```

jsf9k commented on molecule/default/converge.yml:
```diff
@@ -5,3 +5,10 @@
     - name: Include ansible-role-clamav
       ansible.builtin.include_role:
         name: ansible-role-clamav
+      vars:
+        clamav_freshclam_conf:
+          DatabaseMirror: ['db.local.clamav.net', 'database.clamav.net', 'dummy.localhost']
+          Bytecode:
+        clamav_clamd_conf:
+          LogFile: /var/log/clamav/clamav.log
+          StreamMaxLength: 22M
```
This should be in a separate scenario.  I will make this change for you.

jsf9k replied:
Please see commits 048ffd5, 89709c2, 2514295, and 1aadfd5.

jsf9k reviewed (changes requested):
I'll create the second scenario, if you can go ahead and make these alphabetization changes (assuming you're OK with them).

There is still a question as to whether or not these sort of config changes belong in this role.  It may make sense to leave them outside the role, which is the way I am leaning.  (@ClementJ35, of course, is free to fork the role to meet your own needs if we go this route.)  What do you think @dav3r and @mcdonnnj?

jsf9k committed:
Revert change that added vars to the default converge playbook

This change should be tested in a separate scenario.

jsf9k committed:
Move Molecule tests inside the default scenario

jsf9k committed:
Create modify_config scenario

Also move the config modification tests to this scenario.

jsf9k committed:
Add Debian-, Fedora-, and Ubuntu-specific scenarios to test config modification

Also add these new scenarios to the GH Actions build workflow.

mcdonnnj commented on defaults/main.yml:
```diff
@@ -1 +1,20 @@
+#   LogFile: /var/log/clamav/clamav.log
+
+
+clamav_freshclam_conf: false
+clamav_clamd_conf: false
```
If these are dictionaries then the default values should be empty dictionaries.

```suggestion
clamav_freshclam_conf: {}
clamav_clamd_conf: {}
```

jsf9k replied:
Note that this change will impact the where clauses [here](https://github.com/ClementJ35/ansible-role-clamav/blob/1aadfd5d847e9cca24c6301e819a8f925ec1871c/tasks/main.yml#L62-L74).

mcdonnnj replied:
> Note that this change will impact the where clauses [here](https://github.com/ClementJ35/ansible-role-clamav/blob/1aadfd5d847e9cca24c6301e819a8f925ec1871c/tasks/main.yml#L62-L74).

I tested this change locally and ran into no issues. If there's something I missed please let me know.

jsf9k replied:
Ah, OK.  I guess an empty dictionary evaluates to `false` in Ansible.

mcdonnnj commented on README.md:
```diff
@@ -19,6 +19,39 @@ None.
 | Variable | Description | Default | Required |
 |----------|-------------|---------|----------|
 | clamav_cron_frequency | The frequency of ClamAV scanning.  Must be one of: `hourly`, `daily`, `weekly`, or `monthly`. | `weekly` | No |
+| clamav_freshclam_conf | A dictionary of values to set in the freshclam configuration file. | false | No |
+| clamav_clamd_conf | A dictionary of values to set in the clamd configuration file. | false | No |
+| clamav_configuration_backup | Whether or not to backup configuration files before changing. | false | No |
```
These should be sorted. I also changed the default of the two values that are supposed to be dictionaries and made sure all values are code formatted.

```suggestion
| clamav_clamd_conf | A dictionary of values to set in the clamd configuration file. | `{}` | No |
| clamav_configuration_backup | Whether or not to backup configuration files before changing. | `false` | No |
| clamav_cron_frequency | The frequency of ClamAV scanning.  Must be one of: `hourly`, `daily`, `weekly`, or `monthly`. | `weekly` | No |
| clamav_freshclam_conf | A dictionary of values to set in the freshclam configuration file. | `{}` | No |
```

jsf9k replied:
In @mcdonnnj's comment above, "sorted" means "sorted alphabetically".

ClementJ35 committed:
Apply alphabetization and Default as empty dict instead of false

Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>
Co-authored-by: Shane Frasier <maverick@maverickdolphin.com>

ClementJ35 commented:
Hi,
I don't understand the need of alphabetic order. In my opinion, it does make more sense to have the most useful variable first, then go to the least. But I'm fine with that too if you need it.
Did you make your opinion about getting configuration here ? it's not clear to me, while I assume you did because you spent time here.

jsf9k commented:
> I don't understand the need of alphabetic order. In my opinion, it does make more sense to have the most useful variable first, then go to the least. But I'm fine with that too if you need it.

We used to do it that way, but we decided to start alphabetizing because:
1. There can be differences of opinion as to whether variable `x` is more or less useful than variable `y`.
2. When there are more than a handful of variables it is easier to find the one you are interested in if they are listed alphabetically.
3. Alphabetization leads to global consistency across all our Ansible roles.

jsf9k commented:
> Did you make your opinion about getting configuration here ?

We did not.  I didn't hear anything from @mcdonnnj or @dav3r - do either of you two have an opinion?

dav3r commented:
> There is still a question as to whether or not these sort of config changes belong in this role. It may make sense to leave them outside the role, which is the way I am leaning. (@ClementJ35, of course, is free to fork the role to meet your own needs if we go this route.) What do you think @dav3r and @mcdonnnj?

@jsf9k - What did you have in mind for managing this config elsewhere?  How do you feel about including @ClementJ35's config changes in this role for now and then moving to your solution later?

jsf9k commented:
> @jsf9k - What did you have in mind for managing this config elsewhere? How do you feel about including @ClementJ35's config changes in this role for now and then moving to your solution later?

My thought was that it might make more sense to modify the config to your liking using `ansible.builtin.lineinfile`, `ansible.builtin.blockinfile`, etc. as a separate piece of Ansible code, and let this role _only_ handle the installation of clamav (as it did before). 
 Alternatively, the config wrangling could be split off into a separate role.

I'm fine either way; I just wanted to gather opinions from @cisagov/team-ois.

dav3r commented:
> > @jsf9k - What did you have in mind for managing this config elsewhere? How do you feel about including @ClementJ35's config changes in this role for now and then moving to your solution later?
> 
> My thought was that it might make more sense to modify the config to your liking using `ansible.builtin.lineinfile`, `ansible.builtin.blockinfile`, etc. as a separate piece of Ansible code, and let this role _only_ handle the installation of clamav (as it did before). Alternatively, the config wrangling could be split off into a separate role.
> 
> I'm fine either way; I just wanted to gather opinions from @cisagov/team-ois.

I don't have strong feelings either way here, so don't let me stand in the way of progress.

mcdonnnj commented:
> > @jsf9k - What did you have in mind for managing this config elsewhere? How do you feel about including @ClementJ35's config changes in this role for now and then moving to your solution later?
> 
> My thought was that it might make more sense to modify the config to your liking using `ansible.builtin.lineinfile`, `ansible.builtin.blockinfile`, etc. as a separate piece of Ansible code, and let this role _only_ handle the installation of clamav (as it did before). Alternatively, the config wrangling could be split off into a separate role.
> 
> I'm fine either way; I just wanted to gather opinions from @cisagov/team-ois.

I guess I can see it either way. It makes sense to me to want to use a role to install and configure in one swoop if it's for a specific application like this, but I can also see keeping installation and configuration separate. That is all to say that I'm fine with whatever path we go forward with as both have pros/cons attached and I don't believe we have much in the way of precedent/best practice.

jsf9k commented:
> > > @jsf9k - What did you have in mind for managing this config elsewhere? How do you feel about including @ClementJ35's config changes in this role for now and then moving to your solution later?
> > 
> > 
> > My thought was that it might make more sense to modify the config to your liking using `ansible.builtin.lineinfile`, `ansible.builtin.blockinfile`, etc. as a separate piece of Ansible code, and let this role _only_ handle the installation of clamav (as it did before). Alternatively, the config wrangling could be split off into a separate role.
> > I'm fine either way; I just wanted to gather opinions from @cisagov/team-ois.
> 
> I guess I can see it either way. It makes sense to me to want to use a role to install and configure in one swoop if it's for a specific application like this, but I can also see keeping installation and configuration separate. That is all to say that I'm fine with whatever path we go forward with as both have pros/cons attached and I don't believe we have much in the way of precedent/best practice.

OK, let's leave it as it is then.  The config jazz stays.

jsf9k commented:
@ClementJ35 - I took the liberty of removing the last two "ping CI" commits and force pushing.  If the tests pass (as I expect they will) then I'll approve.

Sorry for dropping the ball on this.  I didn't realize that the tests were now passing.

jsf9k reviewed (approved):
Great work!  Thank you for the contribution!

dav3r commented on tasks/configure.yml:
```diff
@@ -0,0 +1,44 @@
+    path: "{{ task_conf_file }}"
+    search_string: "{{ item.key }} "
+    state: present
+  loop: "{{ task_conf_parameters | dict2items | selectattr('value','string') }}"
+
+- name: Configure multi-valued entries in {{ task_conf_file }}
+  ansible.builtin.lineinfile:
+    line: "{{ item.0.key }} {{ item.1 }}"
+    path: "{{ task_conf_file }}"
+    state: present
+  loop: "{{ task_conf_parameters | dict2items | rejectattr('value','string') | rejectattr('value','none') | subelements( 'value' ) }}"
+
+- name: Remove unwanted values from {{ task_conf_file }}
+  ansible.builtin.lineinfile:
+    path: "{{ task_conf_file }}"
+    regexp: "^{{ item.key }} (?!{% for clamav_value in item.value %}{{ clamav_value | regex_escape() }}{{ '|' if not loop.last }}{% endfor %})"
+    state: absent
+  loop: "{{ task_conf_parameters | dict2items | rejectattr('value','none') }}"
+
+- name: Remove unwanted keys from {{ task_conf_file }}
+  ansible.builtin.lineinfile:
+    path: "{{ task_conf_file }}"
+    regexp: "^{{ item.key }} .*$"
+    state: absent
+  loop: "{{ task_conf_parameters | dict2items | selectattr('value','none') }}"
```
Just some minor whitespace formatting to match our usual style:
```suggestion
  loop: "{{ task_conf_parameters | dict2items | selectattr('value', 'string') }}"

- name: Configure multi-valued entries in {{ task_conf_file }}
  ansible.builtin.lineinfile:
    line: "{{ item.0.key }} {{ item.1 }}"
    path: "{{ task_conf_file }}"
    state: present
  loop: "{{ task_conf_parameters | dict2items | rejectattr('value', 'string') | rejectattr('value', 'none') | subelements('value') }}"

- name: Remove unwanted values from {{ task_conf_file }}
  ansible.builtin.lineinfile:
    path: "{{ task_conf_file }}"
    regexp: "^{{ item.key }} (?!{% for clamav_value in item.value %}{{ clamav_value | regex_escape() }}{{ '|' if not loop.last }}{% endfor %})"
    state: absent
  loop: "{{ task_conf_parameters | dict2items | rejectattr('value', 'none') }}"

- name: Remove unwanted keys from {{ task_conf_file }}
  ansible.builtin.lineinfile:
    path: "{{ task_conf_file }}"
    regexp: "^{{ item.key }} .*$"
    state: absent
  loop: "{{ task_conf_parameters | dict2items | selectattr('value', 'none') }}"
```

dav3r reviewed (approved):
This looks excellent to me.  I noted one minor formatting thing, but the rest looks great - thanks for contributing!

jsf9k committed:
Fix up formatting

These changes make the formatting match our usual style.

Co-authored-by: dav3r <daver@geekpad.com>

ClementJ35 commented:
> @ClementJ35 - I took the liberty of removing the last two "ping CI" commits and force pushing. If the tests pass (as I expect they will) then I'll approve.
> 
> Sorry for dropping the ball on this. I didn't realize that the tests were now passing.

Hi @jsf9k, no problem.
Let me know when you merge this.
See you.

mcdonnnj commented on README.md:
```diff
@@ -18,7 +18,40 @@ None.
 
 | Variable | Description | Default | Required |
 |----------|-------------|---------|----------|
+| clamav_clamd_conf | A dictionary of values to set in the clamd configuration file. | `{}` | No |
```
Since we have `clamav_configuration_backup` I would rather use the long form for consistency.

```suggestion
| clamav_clamd_configuration | A dictionary of values to set in the clamd configuration file. | `{}` | No |
```

mcdonnnj commented on README.md:
```diff
@@ -18,7 +18,40 @@ None.
+| clamav_clamd_conf | A dictionary of values to set in the clamd configuration file. | `{}` | No |
+| clamav_configuration_backup | Whether or not to backup configuration files before changing. | `false` | No |
 | clamav_cron_frequency | The frequency of ClamAV scanning.  Must be one of: `hourly`, `daily`, `weekly`, or `monthly`. | `weekly` | No |
+| clamav_freshclam_conf | A dictionary of values to set in the freshclam configuration file. | `{}` | No |
```
Since we have `clamav_configuration_backup` I would rather use the long form for consistency.

```suggestion
| clamav_freshclam_configuration | A dictionary of values to set in the freshclam configuration file. | `{}` | No |
```

mcdonnnj commented on README.md:
````diff
@@ -18,7 +18,40 @@ None.
+### Example ###
+
+```yaml
+clamav_freshclam_conf:
````
With my previous comment applied.

```suggestion
clamav_freshclam_configuration:
```

mcdonnnj commented on defaults/main.yml:
```diff
@@ -1 +1,20 @@
+#   LogFile: /var/log/clamav/clamav.log
+
+
+clamav_freshclam_conf: {}
+clamav_clamd_conf: {}
```
With my previous comments applied.

```suggestion
clamav_freshclam_configuration: {}
clamav_clamd_configuration: {}
```

mcdonnnj commented on molecule/modify_config/converge.yml:
```diff
@@ -0,0 +1,14 @@
+      ansible.builtin.include_role:
+        name: ansible-role-clamav
+      vars:
+        clamav_freshclam_conf:
+          DatabaseMirror: ['db.local.clamav.net', 'database.clamav.net', 'dummy.localhost']
+          Bytecode:
+        clamav_clamd_conf:
```
With my previous comments applied.

```suggestion
        clamav_freshclam_configuration:
          DatabaseMirror: ['db.local.clamav.net', 'database.clamav.net', 'dummy.localhost']
          Bytecode:
        clamav_clamd_configuration:
```

mcdonnnj commented on tasks/main.yml:
```diff
@@ -59,6 +59,20 @@
+  ansible.builtin.include_tasks: configure.yml
+  vars:
+    task_conf_file: "{{ freshclam_configuration_path }}"
+    task_conf_parameters: "{{ clamav_freshclam_conf }}"
+  when: clamav_freshclam_conf
```
With my previous comments applied.

```suggestion
    task_conf_parameters: "{{ clamav_freshclam_configuration }}"
  when: clamav_freshclam_configuration
```

mcdonnnj commented on tasks/main.yml:
```diff
@@ -59,6 +59,20 @@
+  ansible.builtin.include_tasks: configure.yml
+  vars:
+    task_conf_file: "{{ clamd_configuration_path }}"
+    task_conf_parameters: "{{ clamav_clamd_conf }}"
+  when: clamav_clamd_conf
```
With my previous comments applied.

```suggestion
    task_conf_parameters: "{{ clamav_clamd_configuration }}"
  when: clamav_clamd_configuration
```

mcdonnnj commented on defaults/main.yml:
```diff
@@ -1 +1,20 @@
+# A Nonetype value means that the key will be removed from the file.
+# Example:
+#
+# clamav_freshclam_conf:
+#   DatabaseMirror: ['db.local.clamav.net', 'database.clamav.net']
+#   Bytecode: 'true'
+#   PrivateMirror:
+# clamv_clamd_conf:
```
With my previous comments applied. This also happens to fix a typo.

```suggestion
# clamav_freshclam_configuration:
#   DatabaseMirror: ['db.local.clamav.net', 'database.clamav.net']
#   Bytecode: 'true'
#   PrivateMirror:
# clamav_clamd_configuration:
```

mcdonnnj reviewed (approved):
This has polished up nicely. I have a number of suggestions that are just performing the following:
`clamav_freshclam_conf -> clamav_freshclam_configuration`
`clamav_clamd_conf -> clamav_clamd_configuration`

I prefer variable names to avoid shorthand and we already have the `clamav_configuration_backup` variable so it also makes things consistent.

jsf9k committed:
Rename role variable

The new name does not abbreviate "configuration" as "conf" and is therefore clearer.

Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>

jsf9k committed:
Update role variable name in one more place

Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>
